src/optimizer/expression_heuristics.cpp

```diff
@@ -0,0 +1,233 @@
+#include "duckdb/optimizer/expression_heuristics.hpp"
+#include "duckdb/planner/expression/list.hpp"
+
+using namespace duckdb;
+using namespace std;
+
+unique_ptr<LogicalOperator> ExpressionHeuristics::Rewrite(unique_ptr<LogicalOperator> op) {
+    VisitOperator(*op);
+    return op;
+}
+
+void ExpressionHeuristics::VisitOperator(LogicalOperator &op) {
+    if (op.type == LogicalOperatorType::FILTER) {
+        //reorder all filter expressions
+		if (op.expressions.size() > 1) {
+			ReorderExpressions(op.expressions);
+		}
+    }
+
+    //traverse recursively through the operator tree
+    VisitOperatorChildren(op);
+}
+
+void ExpressionHeuristics::ReorderExpressions(vector<unique_ptr<Expression>> &expressions) {
+
+	struct expr {
+		index_t idx;
+		index_t cost;
+
+		bool operator==(const expr &p) const {
+		return idx == p.idx && cost == p.cost;
+		}
+		bool operator<(const expr &p) const {
+		return cost < p.cost || (cost == p.cost && idx < p.idx);
+		}
+	};
+
+	vector<expr> exprCosts;
+	//iterate expressions, get cost for each one and order by that cost
+	for (index_t i = 0; i < expressions.size(); i++) {
+		exprCosts.push_back({i, Cost(*expressions[i])});
+	}
+
+	sort(exprCosts.begin(), exprCosts.end());
+	assert(exprCosts.size() == expressions.size());
+
+	//for all elements to put in place
+	for(index_t i = 0; i < expressions.size() - 1; ++i) {
```

Comment:
Please move the expression into the struct (expr) and replace this nested loop with a single loop placing the expressions back in their correct positions.

Also place rename "expr" to something longer to avoid potential name clashes later.